File: bindings/python/satox_bindings/asset_manager.py
"""
Satox SDK Asset Manager Binding
"""

from typing import Dict, List, Optional, Any
import asyncio
from .error import SatoxError
from .types import Asset, AssetType

class AssetManager:
    """Wrapper for the C++ AssetManager"""
    
    def __init__(self):
        self._core = None
        self._initialized = False
        self._callbacks = []
    
    async def initialize(self, config: Dict[str, Any]) -> bool:
        """Initialize asset manager"""
        if self._initialized:
            raise SatoxError("AssetManager already initialized")
        
        try:
            result = await self._call_native("initialize", config)
            self._initialized = result
            return result
        except Exception as e:
            raise SatoxError(f"Failed to initialize AssetManager: {str(e)}")
    
    async def create_asset(self, config: Dict[str, Any]) -> Asset:
        """Create a new asset"""
        self._ensure_initialized()
        
        try:
            result = await self._call_native("create_asset", config)
            return Asset(**result)
        except Exception as e:
            raise SatoxError(f"Failed to create asset: {str(e)}")
    
    async def get_asset(self, asset_id: str) -> Asset:
        """Get asset information"""
        self._ensure_initialized()
        
        try:
            result = await self._call_native("get_asset", asset_id)
            return Asset(**result)
        except Exception as e:
            raise SatoxError(f"Failed to get asset: {str(e)}")
    
    async def update_asset(self, asset_id: str, metadata: Dict[str, Any]) -> bool:
        """Update asset metadata"""
        self._ensure_initialized()
        
        try:
            return await self._call_native("update_asset", asset_id, metadata)
        except Exception as e:
            raise SatoxError(f"Failed to update asset: {str(e)}")
    
    async def transfer_asset(self, asset_id: str, from_address: str, to_address: str, amount: int) -> str:
        """Transfer asset"""
        self._ensure_initialized()
        
        try:
            return await self._call_native("transfer_asset", asset_id, from_address, to_address, amount)
        except Exception as e:
            raise SatoxError(f"Failed to transfer asset: {str(e)}")
    
    async def get_balance(self, address: str, asset_id: Optional[str] = None) -> Dict[str, int]:
        """Get asset balance"""
        self._ensure_initialized()
        
        try:
            return await self._call_native("get_balance", address, asset_id)
        except Exception as e:
            raise SatoxError(f"Failed to get balance: {str(e)}")
    
    async def list_assets(self, owner: Optional[str] = None, asset_type: Optional[AssetType] = None) -> List[Asset]:
        """List assets"""
        self._ensure_initialized()
        
        try:
            result = await self._call_native("list_assets", owner, asset_type.value if asset_type else None)
            return [Asset(**asset) for asset in result]
        except Exception as e:
            raise SatoxError(f"Failed to list assets: {str(e)}")
    
    def _ensure_initialized(self):
        if not self._initialized:
            raise SatoxError("AssetManager not initialized")
    
    async def _call_native(self, method: str, *args) -> Any:
        """Call native C++ method"""
        raise NotImplementedError("Native binding not implemented") 

File: bindings/typescript/src/asset-manager.ts
/**
 * Satox SDK Asset Manager Binding
 */

import { SatoxError } from './error';
import { Asset, AssetType } from './types';

export class AssetManager {
    private core: any;
    private initialized: boolean = false;
    private callbacks: Array<(event: string, data: any) => void> = [];

    constructor() {
        this.core = require('./native/satox_core');
    }

    async initialize(config: Record<string, any>): Promise<boolean> {
        if (this.initialized) {
            throw new SatoxError('AssetManager already initialized');
        }

        try {
            const result = await this.core.assetInitialize(config);
            this.initialized = result;
            return result;
        } catch (error) {
            throw new SatoxError(`Failed to initialize AssetManager: ${error.message}`);
        }
    }

    async createAsset(config: Record<string, any>): Promise<Asset> {
        this.ensureInitialized();

        try {
            return await this.core.assetCreate(config);
        } catch (error) {
            throw new SatoxError(`Failed to create asset: ${error.message}`);
        }
    }

    async getAsset(assetId: string): Promise<Asset> {
        this.ensureInitialized();

        try {
            return await this.core.assetGet(assetId);
        } catch (error) {
            throw new SatoxError(`Failed to get asset: ${error.message}`);
        }
    }

    async updateAsset(assetId: string, metadata: Record<string, any>): Promise<boolean> {
        this.ensureInitialized();

        try {
            return await this.core.assetUpdate(assetId, metadata);
        } catch (error) {
            throw new SatoxError(`Failed to update asset: ${error.message}`);
        }
    }

    async transferAsset(assetId: string, fromAddress: string, toAddress: string, amount: number): Promise<string> {
        this.ensureInitialized();

        try {
            return await this.core.assetTransfer(assetId, fromAddress, toAddress, amount);
        } catch (error) {
            throw new SatoxError(`Failed to transfer asset: ${error.message}`);
        }
    }

    async getBalance(address: string, assetId?: string): Promise<Record<string, number>> {
        this.ensureInitialized();

        try {
            return await this.core.assetGetBalance(address, assetId);
        } catch (error) {
            throw new SatoxError(`Failed to get balance: ${error.message}`);
        }
    }

    async listAssets(owner?: string, assetType?: AssetType): Promise<Asset[]> {
        this.ensureInitialized();

        try {
            return await this.core.assetList(owner, assetType);
        } catch (error) {
            throw new SatoxError(`Failed to list assets: ${error.message}`);
        }
    }

    registerCallback(callback: (event: string, data: any) => void): void {
        this.callbacks.push(callback);
    }

    unregisterCallback(callback: (event: string, data: any) => void): void {
        const index = this.callbacks.indexOf(callback);
        if (index > -1) {
            this.callbacks.splice(index, 1);
        }
    }

    private ensureInitialized(): void {
        if (!this.initialized) {
            throw new SatoxError('AssetManager not initialized');
        }
    }
} 